### ISSUE TITLE
asyncefile.readLine is unreasonably slow

### ISSUE THREAD

**Issue by skyfex** [nim-lang/Nim#6054]
The asynchronous readLine procedure is many times slower than the synchronous version on Windows (Nim 0.17). This seems to be because it reads data a single character at a time.

**Comment by Araq** [comment]
Yeah, it needs to use an async socket (which is buffered).

**Comment by cheatfate** [comment]
@Araq asyncfile to use async socket?

**Comment by Araq** [comment]
@cheatfate hmm what was I thinking. :-)  But it needs to do buffering.

**Comment by skyfex** [comment]
btw, when I tried to use asyncfile on linux, the read operation never returns, so the program stalls. Is that a known issue, or could it be a problem with the linux distribution (it's CentOS 6.8)? Async httpclient works.

**Comment by cheatfate** [comment]
@Araq yeah implement same mechanism for file, pipe, socket in asyncsocket, asyncfile and asyncpipe... This is way how we are doing?

**Comment by cheatfate** [comment]
@skyfex read operation never returns, so maybe there nothing to read?
Current asyncfile implementation is not using real asynchronous read, but it uses non-blocking read. So if it blocks, before performing real `read()` syscall it uses `epoll()` to poll file descriptor. So it can block only if this file descriptor is not available to read operation.

**Comment by hamidb80** [comment]
**test:**

```nim
import std/[asyncfile, asyncdispatch], times, sequtils

const path = "./play.json"

var f1 = open(path, fmRead)
var f2 = openAsync(path, fmRead)

proc main(): Future[tuple[sync:float, async:float]] {.async.} =
  var acc: seq[(float, float)]
  for _ in 1 .. 1000:
    let s1 = epochTime()
    discard readline f1
    let s2 = epochTime()
    discard await readline f2
    let s3 = epochTime()

    acc.add (s2-s1, s3-s2)

  return acc.foldl (a[0] + b[0], a[1] + b[1])


echo waitfor main()
``

result
```
(sync: 0.003976106643676758, async: 0.4138431549072266)
```

windows 10 64bit
nim 1.6.0
